ide/gototest/src/org/netbeans/api/gototest/TestOppositesLocator.java

```diff
@@ -0,0 +1,194 @@
+                                    if (msg != null) {
+                                        DialogDisplayer.getDefault().notify(
+                                                new NotifyDescriptor.Message(msg, NotifyDescriptor.INFORMATION_MESSAGE));
+                                    }
+                                } else {
+                                    locationResults.put(location, fileObject.getName());
+                                }
+                            }
+                            allDone.countDown();
+                        }
+                    });
+                } else {
+                    TestLocator.LocationResult opposite = locator.findOpposite(fo, caretOffset);
+
+                    if (opposite != null) {
+                        FileObject fileObject = opposite.getFileObject();
+                        if (fileObject == null) {
+                            String msg = opposite.getErrorMessage();
+                            if (msg != null) {
+                                DialogDisplayer.getDefault().notify(
+                                        new NotifyDescriptor.Message(msg, NotifyDescriptor.INFORMATION_MESSAGE));
+                            }
+                        } else {
+                            locationResults.put(opposite, fileObject.getName());
+                        }
+                    }
+
+                    allDone.countDown();
+                }
+            }
+        }
+
+        try {
+            allDone.await();
+        } catch (InterruptedException ex) {
+            Exceptions.printStackTrace(ex);
+        }
+
+        return locationResults.entrySet()
+                              .stream()
+                              .map(e -> new NamedLocation(e.getKey(),
+                                                          e.getValue()))
+                              .collect(Collectors.toList());
+    }
+
+    private TestLocator getLocatorFor(FileObject fo) {
+        Collection<? extends TestLocator> locators = Lookup.getDefault().lookupAll(TestLocator.class);
+        for (TestLocator locator : locators) {
+            if (locator.appliesTo(fo)) {
+                return locator;
+            }
+        }
+
+        return null;
+    }
+
+    private TestLocator.FileType getFileType(FileObject fo) {
+        TestLocator locator = getLocatorFor(fo);
+        if (locator != null) {
+            return locator.getFileType(fo);
+        }
+
+        return TestLocator.FileType.NEITHER;
+    }
+
+    private TestLocator.FileType getCurrentFileType(FileObject fo) {
+        return (fo != null) ? getFileType(fo) : TestLocator.FileType.NEITHER;
+    }
+
+    /**
+     * A description of the found opposite files. Exactly one of {@code errorMessage}
+     * {@code locations} will be non-null;
+     */
+    public static final class LocatorResult {
+        public final String errorMessage;
```

Comment:
I understand that the type is just an immutable tuple/structure ... but shouldn't it have getters rather than allow direct access to fields ?